lib/codeowners/cleaner.rb

```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+require 'git'
+require 'logger'
+
+require_relative 'checker/code_owners'
+require_relative 'checker/file_as_array'
+require_relative 'checker/whitelist'
+
+module Codeowners
+  # Cleans a codeowner file by sorting, grouping and removing
+  # useless patterns, all automatically
+  class Cleaner
+    # Get repo metadata and compare with the owners
+    def initialize(repo)
+      @git = Git.open(repo, log: Logger.new(IO::NULL))
+      @repo_dir = repo
+    end
+
+    def clean!
+      extract_patterns!
+      normalize_patterns!
+      remove_duplicates!
+      remove_useless!
+      group_and_persist!
```

Comment:
I like how the Method reads. 
Did you consider turning this into an object? Something like `PatternCollection` so that we end up with something like:
```ruby
PatternCollection.from_git_ls(some-glob).normalize.uniqu.clean.group.persist?
```

Reply:
This is a really great idea!

Reply:
Actually, what if I just use plain arrays and helpers method, rewriting it like this:

```ruby
    def clean!
      patterns = 
        extract_patterns!
        .map{|p| normalize_pattern(p)}
        .uniq{|p| pattern_id(p)}
        .select{|p| pattern_has_files(p)}

      group_and_persist!(patterns)
    end
```